File: turma.cpp
/*
	AULA DE PARADIGMAS DE PROGRAMAÇÃO
	PROFESSOR: PEDRO CORREA
	ALUNO: PEDRO LUCAS SOARES DA SILVA
	DATA: 25 DE JULHO DE 2024
*/

#include <stdio.h>  // Para interação de entrada e saída com o usuário
#include <locale.h> // Para definir linguagem e acentuação gráfica
#include <string.h> // Para usar o strcmp() -- comparar string
#include <conio.h>  // Para usar o getch() -- segurar a tela
#include <stdlib.h> // Para usar o system.cls(), system("pause");

#define TAM_TURMA 30

struct tipo_aluno {
	char mat[7]; // os que forem do mesmo tipo pode ficar na mesma linha
	char nome[50];
	char sexo;
	float ap1=0;
	float ap2=0;
};

struct tipo_turma {
	char nome[8];
	tipo_aluno alunos[TAM_TURMA];
	int total_al=0;
};

// Reservado para as funções
void cadastrar(tipo_turma *turma){
	
	// TODO: ver se turma já não está cheia
	
	fflush(stdin);
	printf("\nDigite a matrícula do aluno: ");
	gets(turma->alunos[turma->total_al].mat);
	
	fflush(stdin);
	printf("\nDigite o nome do aluno: ");
	gets(turma->alunos[turma->total_al].nome);
	
	fflush(stdin);
	printf("\nDigite o sexo do aluno [M/F]: ");
	turma->alunos[turma->total_al].sexo = getchar();
	
	printf("\nAluno \"%s\" cadastrado com sucesso!", turma->alunos[turma->total_al].nome);
	turma->total_al++;
	
}

int main(){
	
	setlocale(LC_ALL,"Portuguese");
	
	tipo_turma turma;
	int op;
	
	while(1){
		system("cls");
		printf("\n======== MENU ========");
		printf("\n [1] Cadastrar aluno");
		printf("\n [2] Listar turma");
		printf("\n [3] Pesquisar aluno");
		printf("\n [4] Inserir nota");
		printf("\n [5] Mostrar aprovados");
		printf("\n [0] Sair");
//		printf("\n [ ] <mock>");
		
		printf("\nDigite sua opção: ");
		scanf("%d", &op);
		
		printf("\n[DEBUG]: Início do retorno:\n");
		switch(op){
			case 0:
				return 0;
//				break; // [not needed]
			case 1:
				cadastrar(&turma);
				break;
			case 2:
//				listar(&turma);
				break;
			case 3:
//				pesquisar(&turma);
				break;
			case 4:
//				inserir(&turma);
				break;
			case 5:
//				aprovados(&turma);
				break;
			default:
				printf("\nOpção inválida!\n");
		};
		
		printf("\n[DEBUG]: Fim do retorno.\n");
		getch();
		
	};
	
	return 0;
		
}

File: turma.pas
program Hello;
uses crt;
{This language is not case sensitive}

const TAM = 30;
type
tipo_aluno = record
    mat : array[1..7] of char;
    nome: array[1..50] of char;
    sexo: char;
    p1: real;
    p2: real;
    end;

{Reservado para declarar o ponteiro da turma}
tipo_ponteiro = ^tipo_turma;

tipo_turma = record
    nome: array[1..20] of char;
    alunos: array[1..TAM] of tipo_aluno;
    total_al: integer;
    end;

{a partir daqui serão declarados os procedimentos}
procedure cadastrar(turma: tipo_ponteiro);
var
    sexo: char;
begin
    clrscr;
    if (turma^.total_al = TAM) then
        begin
            writeln;
            writeln('* IMPOSSÍVEL CADASTRAR! Motivo: A turma está completa!');
            writeln('* PRESSIONE QUALQUER TECLA PARA VOLTAR AO MENU.          *');
            writeln;
        end
    else
        begin
        	clrscr; { limpa a tela } 
        	writeln(); writeln();
        	writeln('**********************************************************');		
        	writeln('*                       MATRICULANDO ALUNO               *');	
        	writeln('**********************************************************');

            turma^.total_al := turma^.total_al + 1;

            write('Digite a matrícula: ');
            readln(turma^.alunos[turma^.total_al].mat);
            write('Digite o nome: ');
            readln(turma^.alunos[turma^.total_al].nome);
            write('Digite o sexo: ');
            readln(sexo); {armazena em variável separada}
            turma^.alunos[turma^.total_al].sexo := upcase(sexo); {converte em maiúscula para padronizar}
            writeln;writeln;
            writeln('   ALUNO CADASTRADO COM SUCESSO !!!');
        end;
end;

procedure listar(turma: tipo_ponteiro);
var
    sexo: char;
begin
    clrscr;
    if (turma^.total_al = TAM) then
        begin
            writeln;
            writeln('* IMPOSSÍVEL LISTAR! Motivo: A turma está vazia.');
            writeln('* PRESSIONE QUALQUER TECLA PARA VOLTAR AO MENU.          *');
            writeln;
        end
    else
        begin
        	clrscr; { limpa a tela } 
        	writeln(); writeln();
        	writeln('**********************************************************');		
        	writeln('*                       LISTA DE ALUNOS                  *');	
        	writeln('**********************************************************');
        end;
end;

{Início do programa}
var
    turma: tipo_turma;
    op: integer;

begin

    writeln ('******************************');
    writeln ('       INÍCIO DO PROGRAMA');
    writeln ('Digite o nome da turma:');
    readln(turma.nome);
    turma.total_al := 0;

    while(true) do
        begin
            clrscr;

            writeln ('**********************************************************');
            writeln ('                  = TURMA  = ', turma.nome);
	        writeln ('                   ==== MENU ====');
	        writeln ('                  [1] - Cadastrar aluno');
	        writeln ('                  [2] - Listar turma');
	        writeln ('                  [3] - Pesquisar aluno');
	        writeln ('                  [4] - Cadastrar nota');
	        writeln ('                  [5] - Alterar dados de aluno');
	        writeln ('                  [0] - Sair');
	        writeln ('**********************************************************');
            writeln;writeln;
	        write('                   Digite sua opção: ');
            readln(op);

            case (op) of
                1: cadastrar(@turma);
                2: listar(@turma);
                3: writeln(' pesquisar(@turma);');
                4: writeln(' Cadastrar nota');
                5: writeln(' Alterar dados');
                0: Exit;
            end;

            readln;
        end;
end.